node/src/main/scala/coop/rchain/node/api/WebApi.scala

```diff
@@ -0,0 +1,222 @@
+package coop.rchain.node.api
+
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import com.google.protobuf.ByteString
+import coop.rchain.blockstorage.BlockStore
+import coop.rchain.casper.SafetyOracle
+import coop.rchain.casper.api.BlockAPI
+import coop.rchain.casper.engine.EngineCell.EngineCell
+import coop.rchain.casper.protocol.{BlockInfo, DataWithBlockInfo, DeployData, LightBlockInfo}
+import coop.rchain.crypto.PublicKey
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.crypto.signatures.{SignaturesAlg, Signed}
+import coop.rchain.metrics.Span
+import coop.rchain.models.GUnforgeable.UnfInstance.{GDeployIdBody, GDeployerIdBody, GPrivateBody}
+import coop.rchain.models._
+import coop.rchain.node.api.WebApi._
+import coop.rchain.shared.Log
+
+import scala.collection._
+
+trait WebApi[F[_]] {
+  def status: F[ApiStatus]
+
+  // Get data to create deploy
+  def prepareDeploy(request: Option[PrepareRequest]): F[PrepareResponse]
+
+  // Write data (deploy)
+  def deploy(request: Signed[DeployData]): F[String]
+
+  // Read data (listen)
+  def listenForDataAtName(request: DataRequest): F[DataResponse]
+
+  // Blocks info
+
+  def lastFinalizedBlock: F[BlockInfo]
+
+  def getBlock(hash: String): F[BlockInfo]
+
+  def getBlocks(depth: Option[Int]): F[List[LightBlockInfo]]
+}
+
+object WebApi {
+
+  class WebApiImpl[F[_]: Sync: Concurrent: EngineCell: Log: Span: SafetyOracle: BlockStore]
+      extends WebApi[F] {
+    import WebApiSyntax._
+
+    def prepareDeploy(req: Option[PrepareRequest]): F[PrepareResponse] = {
+      val blockNumber = BlockAPI
+        .getBlocks[F](1.some)
+        .flatMap(_.leftToError)
+        .map(_.headOption.map(_.blockNumber).getOrElse(-1L))
+
+      val previewNames = req.fold(List.empty[String].pure) { r =>
+        BlockAPI
+          .previewPrivateNames[F](toByteString(r.deployer), r.timestamp, r.nameQty)
+          .flatMap(_.leftToError)
+          .map(_.map(toHex).toList)
+      }
+
+      previewNames.map2(blockNumber)(PrepareResponse)
+    }
+
+    def deploy(request: Signed[DeployData]): F[String] =
+      BlockAPI.deploy(request).flatMap(_.leftToError)
```

Comment:
@marcin-rzeznicki when implementing WebAPI is where I'd like implicit conversion so that `.flatMap(_.leftToError)` can be omitted.

Reply:
I don't think that implicit conversion is warrant here. I believe it's better to be explicit

Reply:
Maybe extension method for `.flatMap(_.leftToError)` but that's for another PR.